### PR TITLE
feat(TDI-45590): tMongoDB drivers and API update

### PR DESCRIPTION
**Pull request by dmytrogrygorenko**
**Please check if the PR fulfills these requirements**

- [X] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [ ] Unit tests for the Java changes have been added (for bug fixes / features) ?
- [ ] TUJ for the JavaJet changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [X] The new code does not introduce new technical issues

**What is the current behavior?** (You can also link to an open issue here)
https://jira.talendforge.org/browse/TDI-45590

**What is the new behavior?**
Update tMongoDB* components family to support new MongoDB 4.4 API.

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBInput/tMongoDBInput_begin.javajet [hunk 1]
```diff
@@ -488,7 +331,7 @@
 <%
 						}else if(javaType == JavaTypesManager.BYTE_ARRAY) {
 %>
-						<%if(version_3_5_X) {%>
+						<%if(version_3_5_X || version_4_4_X) {%>
```
Comment (mbasiuk-talend) [inline main/plugins/org.talend.designer.components.bigdata/components/tMongoDBInput/tMongoDBInput_begin.javajet:334]: Very weird to see the same condition many times. Maybe You should extract a single variable?

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBRow/tMongoDBRow_main_below_4_4_X.javajet [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		%>
+		if(result_<%=cid%>.ok()){
+			<%
+			if(isLog4jEnabled){
+			%>
+				log.info("<%=cid%> - Return the result is : "+result_<%=cid%>);
+			<%
+			} else {
+			%>
+			    System.out.println(result_<%=cid%>);
```
Comment (mbasiuk-talend) [inline main/plugins/org.talend.designer.components.bigdata/components/tMongoDBRow/tMongoDBRow_main_below_4_4_X.javajet:64]: Do we really need to use **system out println** in code?

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBWriteConf/GeneralConnection_4_4_X.javajet [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<%@ jet
```
Comment (mbasiuk-talend) [inline main/plugins/org.talend.designer.components.bigdata/components/tMongoDBWriteConf/GeneralConnection_4_4_X.javajet:1]: Why do we need this file?

File: main/plugins/org.talend.designer.components.bigdata/components/templates/MongoDB/GeneralMongoDBConnectionSettings_4_4_X.javajet [hunk 4]
```diff
@@ -0,0 +1,55 @@
+                // SSL
```
Comment (mbasiuk-talend) [inline main/plugins/org.talend.designer.components.bigdata/components/templates/MongoDB/GeneralMongoDBConnectionSettings_4_4_X.javajet:1]: Why doesn't this contain **jet** beginning? I didn't work with such files.

File: main/plugins/org.talend.designer.components.bigdata/components/templates/MongoDB/GridFS/GetMongoDBConnection_below_4_4_X.javajet [hunk 5]
```diff
@@ -0,0 +1,38 @@
+<%@ jet 
+    imports="
+        java.util.Map
+        java.util.List
+    " 
+%>
+	<%@ include file="@{org.talend.designer.components.bigdata}/components/templates/MongoDB/GeneralMongoDBConnectionSettings_below_4_4_X.javajet"%>
+
+			<%if (setReadPreference) {
+			        if (readPreference.equals("PRIMARY")) {
```
Comment (mbasiuk-talend) [inline main/plugins/org.talend.designer.components.bigdata/components/templates/MongoDB/GridFS/GetMongoDBConnection_below_4_4_X.javajet:10]: switch case?

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_begin_4_4_X.javajet [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+    // Declarations
+    com.mongodb.client.MongoClient mongo_<%=cid%>=null;
+    com.mongodb.client.MongoDatabase db_<%=cid%>=null;
+    List<com.mongodb.ServerAddress> addrs_<%=cid%> = new java.util.ArrayList<>();
+    com.mongodb.MongoClientSettings.Builder clientSettingsBuilder_<%=cid%> = com.mongodb.MongoClientSettings.builder();
+    com.mongodb.connection.ClusterSettings.Builder clusterSettingsBuilder_<%=cid%> = com.mongodb.connection.ClusterSettings.builder();
+    com.mongodb.connection.ConnectionPoolSettings.Builder connPoolSettingsBuilder_<%=cid%> = com.mongodb.connection.ConnectionPoolSettings.builder();
+    com.mongodb.connection.ServerSettings.Builder serverSettingsBuilder_<%=cid%> = com.mongodb.connection.ServerSettings.builder();
+    com.mongodb.connection.SocketSettings.Builder sockerSettingsBuilder_<%=cid%> = com.mongodb.connection.SocketSettings.builder();
```
Comment (mbasiuk-talend) [inline main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_begin_4_4_X.javajet:23]: typo in the property name.

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_messages_en.properties [hunk 7]
```diff
@@ -64,3 +65,9 @@ OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.SOCKET_KEEP_ALIVE=socketK
 OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.SOCKET_TIMEOUT=socketTimeout
 OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.SSL_INVALID_HOSTNAME_ALLOWED=sslInvalidHostNameAllowed
 OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.THREADS_ALLOWED_TO_BLOCK_FOR_CONNECTION_MULTIPLIER=threadsAllowedToBlockForConnectionMultiplier
+OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.MAINTENANCE_FREQUENCY = maintenanceFrequency
```
[the remark this answers is not in the document]
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_messages_en.properties]: Done.

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_messages.properties [hunk 8]
```diff
@@ -73,4 +74,10 @@ OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.SERVER_SELECTION_TIMEOUT
 OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.SOCKET_KEEP_ALIVE = socketKeepAlive
 OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.SOCKET_TIMEOUT = socketTimeout
 OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.SSL_INVALID_HOSTNAME_ALLOWED = sslInvalidHostNameAllowed
-OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.THREADS_ALLOWED_TO_BLOCK_FOR_CONNECTION_MULTIPLIER = threadsAllowedToBlockForConnectionMultiplier
\ No newline at end of file
+OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.THREADS_ALLOWED_TO_BLOCK_FOR_CONNECTION_MULTIPLIER = threadsAllowedToBlockForConnectionMultiplier
+OPTIONAL_ATTRIBUTES.ITEM.OPTIONAL_ATTRIBUTES_NAME.ITEM.MAINTENANCE_FREQUENCY = maintenanceFrequency
```
[the remark this answers is not in the document]
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_messages.properties]: Done.

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBGridFSProperties/tMongoDBGridFSProperties_begin.javajet [hunk 9]
```diff
@@ -83,7 +83,7 @@
 							valueObj_<%=cid%> = file.getLength();
 							<%} else if ("contentType".equalsIgnoreCase(columnName)){%>
 							valueObj_<%=cid%> = file.getContentType();
-							<%} else if ("md5".equalsIgnoreCase(columnName)){%>
+							<%} else if (("md5".equalsIgnoreCase(columnName)) && version_3_5_X){%>
```
[the remark this answers is not in the document]
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBGridFSProperties/tMongoDBGridFSProperties_begin.javajet:86]: Not supported. Here's the javadoc:
https://mongodb.github.io/mongo-java-driver/3.12/javadoc/com/mongodb/client/gridfs/model/GridFSFile.html   - 3.12
http://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/client/gridfs/model/GridFSFile.html  - 4.2

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBGridFSProperties/tMongoDBGridFSProperties_messages_en.properties [hunk 10]
```diff
@@ -22,6 +22,8 @@ REPLICA_SET.NAME=Server addresses
 REPLICA_SET.ITEM.REPLICA_HOST=Host
 REPLICA_SET.ITEM.REPLICA_PORT=Port
 DB_VERSION.NAME=DB Version
+DB_VERSION.ITEM.MONGODB_4_4_X=MongoDB 4.4.X and later
```
[the remark this answers is not in the document]
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBGridFSProperties/tMongoDBGridFSProperties_messages_en.properties]: Fixed.

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBInput/tMongoDBInput_begin.javajet [hunk 11]
```diff
@@ -75,189 +79,26 @@
...
-		        <%
-    		}
-
-			if(isLog4jEnabled){
-			    %>	
-				log.info("<%=cid%> - Connecting to " + mongo_<%=cid%>.getServerAddressList() + ".");
-				<%
-			}
-			log4jFileUtil.startRetriveDataInfo();
+            <%if ("MONGODB_4_4_X".equals(dbversion)) {%>
```
[the remark this answers is not in the document]
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBInput/tMongoDBInput_begin.javajet]: Fixed.

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBWriteConf/GeneralConnection_4_4_X.javajet (hunk 3 above)
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBWriteConf/GeneralConnection_4_4_X.javajet]: Forgotten leftover from refactoring, removed now.

File: main/plugins/org.talend.designer.components.bigdata/components/templates/MongoDB/GeneralMongoDBConnectionSettings_4_4_X.javajet (hunk 4 above)
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/templates/MongoDB/GeneralMongoDBConnectionSettings_4_4_X.javajet:7]: Fixed in all templates.

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_begin_4_4_X.javajet (hunk 6 above)
Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBConnection/tMongoDBConnection_begin_4_4_X.javajet]: Fixed.

File: main/plugins/org.talend.designer.components.bigdata/components/tMongoDBOutput/tMongoDBOutput_begin_4_4_X.javajet [hunk 12]
```diff
@@ -112,10 +112,13 @@
             }
             if(dropExistCollection) {%>
                 com.mongodb.client.MongoIterable<String> collectionNames = db_<%=cid%>.listCollectionNames();
-                for (String name : collectionNames) {
-                    if (<%=collection%>.equalsIgnoreCase(name)) {
-                        db_<%=cid%>.getCollection(<%=collection%>).drop();
-                        break;
+                try(com.mongodb.client.MongoCursor<String> iterator_<%=cid%> = collectionNames.iterator()) {
```
Comment (mbasiuk-talend) [inline main/plugins/org.talend.designer.components.bigdata/components/tMongoDBOutput/tMongoDBOutput_begin_4_4_X.javajet:115]: What if we don't use iterable to drop collection?
Have you checked this article?
https://stackoverflow.com/questions/19498650/how-to-drop-collection-from-database-in-mongodb-using-mongo-db-java-driver

Comment (dmytrogrygorenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tMongoDBOutput/tMongoDBOutput_begin_4_4_X.javajet]: Collection drop was simplified.

**Review by mbasiuk-talend — DISMISSED** [review]
LGTM

**Review by mbasiuk-talend — APPROVED** [review]

**Review by mbasiuk-talend — DISMISSED** [review]

**Review by mbasiuk-talend — APPROVED** [review]

**Pull request merged**